src/formatDate.js

```diff
@@ -8,7 +8,68 @@
  * @returns {string}
  */
 function formatDate(date, fromFormat, toFormat) {
-  // write code here
+  if (typeof date !== 'string' || date.length === 0) {
+    return '';
+  }
+
+  if (
+    !fromFormat ||
+    !toFormat ||
+    typeof fromFormat !== 'object' ||
+    typeof toFormat !== 'object' ||
+    !('length' in fromFormat) ||
+    !('length' in toFormat)
+  ) {
+    return '';
+  }
+
+  const fromSeparator = fromFormat[3];
+  const toSeparator = toFormat[3];
+
+  if (!fromSeparator || !toSeparator) {
+    return '';
+  }
+
+  const parts = date.split(fromSeparator);
+
+  if (parts.length !== 3) {
+    return '';
+  }
+
+  const partsMap = {
+    DD: '',
+    MM: '',
+    YYYY: '',
+    YY: '',
+  };
+
+  fromFormat.slice(0, 3).forEach((part, index) => {
+    if (partsMap.hasOwnProperty(part)) {
+      partsMap[part] = parts[index];
+    }
+  });
+
+  if (partsMap['YY']) {
+    const yy = +partsMap['YY'];
+
+    if (yy < 30) {
+      partsMap['YYYY'] = '20' + partsMap['YY'];
+    } else {
+      partsMap['YYYY'] = '19' + partsMap['YY'];
+    }
```

Comment:
The conversion of `YY` to `YYYY` is well-implemented. However, you could optimize this by only performing the conversion if `toFormat` includes `YYYY` and `fromFormat` includes `YY`, since the conversion might not be needed in all cases.